# Capstone project Week 3 assignment

##  Part 3
#### Includes parts 1 and 2

### Import required libraries

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge lxml --yes 
!pip install et_xmlfile

#!conda install -c conda-forge folium=0.11.0 --yes #installed 19jun20
import folium # plotting library


Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.



#### Read the table from the Wikipedia page directly by pandas

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wikitable = pd.read_html(url)

wikitable

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                           Neighborhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

#### It can be seen from the html output that the postal code table is in the [0] position.
#### A pandas data frame is created from position [0] of wikitable. 

In [3]:
df_postalCode=pd.DataFrame(wikitable[0])

df_postalCode

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

#### Remove the boroughs that are "Not Assigned" and check if there will be any "NaN" neighborhood remaining.

In [5]:
notAssignedB = df_postalCode['Borough'] == 'Not assigned'
notAssignedN = df_postalCode['Neighborhood'] == 'Not assigned'
notAssignedIndexB= df_postalCode[notAssignedB].index
notAssignedIndexN= df_postalCode[notAssignedN].index
if all(notAssignedIndexB == notAssignedIndexN):
    df_postalCode.drop(notAssignedIndexB, inplace=True)
    print("None unassigned neighborhoods remain")

None unassigned neighborhoods remain


#### There are no neighborhoods to be renamed after removing the "Not Assigned" boroughs.

In [6]:
print(df_postalCode.shape)
df_postalCode

(103, 3)


Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

#### The resulting data frame has 103 postal codes (rows).
#### Checking if any Postal Code is repeated; if yes, the corresponding neighborhoods will need to be merged.

In [7]:
df_postalCode.nunique()

Postal Code     103
Borough          10
Neighborhood     99
dtype: int64

#### There are 103 different postal codes, which means there are no repeats.
#### There are ten different boroughs in the 103 postal codes. It was known in advance that boroughs are repeated.
#### There are four repeated neighborhoods. Which ones?

In [8]:
dups=df_postalCode['Neighborhood'].duplicated()
df_postalCode[dups].sort_values(by=df_postalCode.columns.to_list())

Postal Code     Borough Neighborhood
20         M3C  North York    Don Mills
74         M3L  North York    Downsview
83         M3M  North York    Downsview
92         M3N  North York    Downsview

In [9]:
dups2 = df_postalCode['Neighborhood'] == 'Don Mills'
dups3 = df_postalCode['Neighborhood'] == 'Downsview'
print(df_postalCode[dups2].sort_values(by=df_postalCode.columns.to_list()),'\n')
print(df_postalCode[dups3].sort_values(by=df_postalCode.columns.to_list()))

   Postal Code     Borough Neighborhood
11         M3B  North York    Don Mills
20         M3C  North York    Don Mills 

   Postal Code     Borough Neighborhood
65         M3K  North York    Downsview
74         M3L  North York    Downsview
83         M3M  North York    Downsview
92         M3N  North York    Downsview


#### According to the repeats lists, Don Mills appears twice in the data frame and Downsview appears four times.
#### However, there is indeed a different postal code for every repeated neighborhood.
#### This means that there is nothing more to clean up in this data set and there are 103 different postal codes in Toronto with an assigned borough and neighborhood.

In [10]:
print(df_postalCode.shape)

df_postalCode

(103, 3)


Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## Part 2

#### Install and Import required library

In [11]:
!pip install pgeocode
import pgeocode

#### Set country to query as Canada ('ca').
#### Create a list 'l_postalCode' of the postal codes from the 'df_postalCode' data frame.
#### Loop over 'l_postalCode' to obtain the geo data for each postal code as a data frame 'df_latlon'.
#### Check the columns of 'df_latlon' to identify where the latitute and logitude are.

In [12]:
nomi = pgeocode.Nominatim('ca')
l_postalCode = df_postalCode['Postal Code'].tolist()
df_latlon = pd.DataFrame()
for i in l_postalCode:
    df_latlon = df_latlon.append (nomi.query_postal_code([i]))

df_latlon.columns

Index(['postal_code', 'country code', 'place_name', 'state_name', 'state_code',
       'county_name', 'county_code', 'community_name', 'community_code',
       'latitude', 'longitude', 'accuracy'],
      dtype='object')

#### Latitiude and Longitude are in columns 'latitude' and 'longitude' of 'df_latlon' respectively.
#### Columns for latitude and longitude are added to 'df_postalCode' data frame from 'df_latlon'.
#### This completes the data frame format required.

In [13]:
df_postalCode['latitude'] = df_latlon['latitude'].tolist()
df_postalCode['longitude'] = df_latlon['longitude'].tolist()

df_postalCode

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  latitude  longitude  
2                                            Parkwoods   43.7545   -79.3300  
3                                     Victoria Village   43.7276   -79.3148  
4                            Regent Park, Harbourfront   43.6555   -79.3626  
5                     Lawrence Manor, Lawrence Heights   43.7223   -79.4504  
6          Queen's Park, Ontario Provincial Government   43.6641   -79.3889  
..                                                 ...       ...        ...  
160      The Kingsway, Montgomery Road, Old Mill North   43.6518   -79.5076  
165                               Church and Wellesley   43.6656   -79.3830  
168  Business reply mail Processing Centre, South C...   43.7804   -79.2505  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...   43.6325   -79.4939  
178  Mimico NW, The Queensway West, South of Bloor,...   43.6256   -79.5231  

[103 rows x 5 columns]

### Check that there is no missing values for latitude and/or longitude.

In [14]:
NAlon = df_postalCode['longitude'].astype('str') == 'nan'
NAlat = df_postalCode['latitude'].astype('str') == 'nan'
NAlonIndex = df_postalCode[NAlon].index
NAlatIndex = df_postalCode[NAlat].index
if all(NAlonIndex == NAlatIndex):
    print(df_postalCode.loc[NAlatIndex])
else:
    print(df_postalCode.loc(NAlatIndex))
    print(df_postalCode.loc(NAlonIndex))

    Postal Code      Borough                           Neighborhood  latitude  \
114         M7R  Mississauga  Canada Post Gateway Processing Centre       NaN   

     longitude  
114        NaN  


### The borough Mississauga with postal code M7R does not have the latitude-longitude information.
### This information will be obtained using geopy with the borough as address.

In [15]:
address = 'Mississauga, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude2 = location.latitude
longitude2 = location.longitude
print('The coordinates of Mississauga are {}, {}.'.format(latitude2, longitude2))

The coordinates of Mississauga are 43.590338, -79.645729.


### These will be added to the data frame 'df_postalCode' at the corresponding position (114).

In [16]:
df_postalCode.loc[NAlatIndex, 'latitude'] = latitude2
df_postalCode.loc[NAlatIndex, 'longitude'] = longitude2

df_postalCode.loc[NAlonIndex]

Postal Code      Borough                           Neighborhood  \
114         M7R  Mississauga  Canada Post Gateway Processing Centre   

      latitude  longitude  
114  43.590338 -79.645729

### The data frame is now complete (no missing values) in the required format.

In [17]:
df_postalCode

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  latitude  longitude  
2                                            Parkwoods   43.7545   -79.3300  
3                                     Victoria Village   43.7276   -79.3148  
4                            Regent Park, Harbourfront   43.6555   -79.3626  
5                     Lawrence Manor, Lawrence Heights   43.7223   -79.4504  
6          Queen's Park, Ontario Provincial Government   43.6641   -79.3889  
..                                                 ...       ...        ...  
160      The Kingsway, Montgomery Road, Old Mill North   43.6518   -79.5076  
165                               Church and Wellesley   43.6656   -79.3830  
168  Business reply mail Processing Centre, South C...   43.7804   -79.2505  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...   43.6325   -79.4939  
178  Mimico NW, The Queensway West, South of Bloor,...   43.6256   -79.5231  

[103 rows x 5 columns]

## Part 3

#### Import required library

In [18]:
from sklearn.cluster import KMeans 
from sklearn.datasets.samples_generator import make_blobs

### Obtain the latitude and logitude values of Toronto.
### A map centered in Toronto will be created using these values.

In [19]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


### Each post code in the data frame 'df_postalCode' will be indicated in the map.

In [20]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhood in zip(df_postalCode['latitude'], df_postalCode['longitude'], df_postalCode['Borough'], df_postalCode['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Kmeans clustering will be used with the latitude and longitude coordinates from the 'df_postalCode' dataframe.
### Five clusters will be defined in order to identify the center of the "main" geographical locations in Toronto.
### For example, the coordinates for Downtown Toronto will be approximated as one of the clusters.
### Other four clusters will be identified and the expectation is to be able to label them in a logical manner, for example, east, west, etc.

In [21]:
kclusters = 5

df_postalCodeCluster = df_postalCode.drop(['Postal Code', 'Borough','Neighborhood'], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(df_postalCodeCluster)

# check cluster centrids
kmeans.cluster_centers_

array([[ 43.73042857, -79.45004286],
       [ 43.780475  , -79.22726667],
       [ 43.66898544, -79.53987383],
       [ 43.6622871 , -79.39097097],
       [ 43.7332381 , -79.32459524]])

### It is necessary to identify where each of these "regions" are located.
### Additionally, a temporary label will be added, from 0 to 4, so that these can be traced back to each of the clusters' coordinates.
### So that according to what it is observed in the map, each cluster can be labeled properly as a region.

In [22]:
for reg, lat, lon in zip(range(5), kmeans.cluster_centers_[:,0], kmeans.cluster_centers_[:,1]):
    label = '{}, {}, {}'.format(reg, lat, lon)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=1.0,
        parse_html=False).add_to(map_toronto) 

In [33]:
map_toronto

### As it was expected, by looking at the map, the red indicators show five different region that can logically be labeled as Downtown, West, Mid west, Mid east and East.
### To identify which coordinates belong to each of the newly labeled regions, we navigate the map and check the first element of the pop up label.
### Let's create a data frame with the information.

In [31]:
regionsTor = pd.DataFrame()

regionsLabel = ['Mid west', 'East','West','Downtown','Mid East']

regionsTor['Region'] = regionsLabel
regionsTor['latitude'] = kmeans.cluster_centers_[:,0]
regionsTor['longitude'] = kmeans.cluster_centers_[:,1]

In [32]:
regionsTor

Region   latitude  longitude
0  Mid west  43.730429 -79.450043
1      East  43.780475 -79.227267
2      West  43.668985 -79.539874
3  Downtown  43.662287 -79.390971
4  Mid East  43.733238 -79.324595

### Let's update the folium map pop ups so that the region label is now shown.

In [34]:
for reg, lat, lon in zip(regionsTor['Region'], regionsTor['latitude'], regionsTor['longitude']):
    label = '{}, {}, {}'.format(reg, lat, lon)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='black',
        fill=True,
        fill_color='red',
        fill_opacity=0.9,
        parse_html=False).add_to(map_toronto) 

In [36]:
map_toronto